### PR TITLE
Private registries support

### PR DESCRIPTION
**Pull request by davidfestal**
### What does this PR do?

This PR adds support of private docker registries when running on the Kubernetes and OpenShift infrastructure.

### What issues does this PR fix or reference?

https://github.com/redhat-developer/rh-che/issues/680

### CODE REVIEW

**Comment by davidfestal** [comment]
ci-test

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                  }
+                  jsonWriter.name(name);
+                  jsonWriter.beginObject();
+                  jsonWriter.name("username");
+                  jsonWriter.value(authConfig.getUsername());
+                  jsonWriter.name("password");
+                  jsonWriter.value(authConfig.getPassword());
+                  jsonWriter.name("email");
+                  jsonWriter.value("email@email");
+                  String auth =
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: Can you point me to a doc that describes how this object should look like - I don't understand by looking at the code. This also indicates that this code is not clear and might require additional code simplifications or comments to ease support of this code.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+          .inNamespace(namespaceFactory.create(identity.getWorkspaceId()).getName())
+          .createOrReplace(secret);
+    } catch (Exception e) {
+      throw new InfrastructureException(e);
+    }
+
+    k8sEnv.getPods().values().forEach(this::provision);
+  }
+
+  public void provision(Pod pod) {
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: It looks like this method should be private

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisionerTest.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        .encodeToString(
+            new StringBuilder()
+                .append(username)
+                .append(':')
+                .append(password)
+                .toString()
+                .getBytes());
+  }
+
+  @SuppressWarnings("unchecked")
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisionerTest.java]: If you use suppressing of warnings, please, add a comment to the doc that explains why it is safe to suppress warnings.

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisionerTest.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+  public void dontDoAnythingIfNoPrivateRegistries() throws Exception {
+    when(authConfigs.getConfigs()).thenReturn(Collections.emptyMap());
+
+    imagePullSecretProvisioner.provision(k8sEnv, runtimeIdentity);
+
+    verifyZeroInteractions(namespaceOperation);
+    verifyZeroInteractions(podSpec);
+  }
+
+  private static class TestAuthConfig implements AuthConfig {
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisionerTest.java]: Please, move auxiliary methods down. It is easier to read test when test cases are higher and all other methods are lower in the class.

**Review by garagatyi — COMMENTED** [review]
Overall looks good. I would love to see code been improved from the support side. I added inlined comments about that.

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisionerTest.java (hunk 3 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisionerTest.java]: BTW it's not good practice to use SuppressWarnings for the whole method.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  @Override
+  public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+
+    AuthConfigs authConfigs = userSpecificDockerRegistryCredentialsProvider.getCredentials();
+    if (authConfigs.getConfigs().isEmpty()) {
+      return;
+    }
+
+    Gson gson = new Gson();
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: Please explain why do you need to create it here instead of a single line where it is used?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+    Base64.Encoder encoder = Base64.getEncoder();
+
+    String config;
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: Why do you need a declaration of this variable here?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    try (StringWriter strWriter = new StringWriter();
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: Maybe private method could be enough

File: infrastructures/kubernetes/pom.xml [hunk 8]
```diff
@@ -114,6 +114,10 @@
             <groupId>org.eclipse.che.core</groupId>
             <artifactId>che-core-db</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.eclipse.che.infrastructure.docker</groupId>
+            <artifactId>docker-client</artifactId>
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/pom.xml]: I think this dependency can potentially bring some issues in future. I suggest to move docker auth related code in separate maven module and replace this dependency with it.
@sleshchenko @garagatyi @gazarenkov wdyt?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                  String auth =
+                      new StringBuilder(authConfig.getUsername())
+                          .append(':')
+                          .append(authConfig.getPassword())
+                          .toString();
+                  jsonWriter.name("auth");
+                  jsonWriter.value(encoder.encodeToString(auth.getBytes()));
+                  jsonWriter.endObject();
+                } catch (IOException e) {
+                  throw new RuntimeException(e);
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: Please don't throw `RuntimeException`. `InfrastructureException` only is expected to be thrown by this method.

File: infrastructures/kubernetes/pom.xml (hunk 8 above)
Comment (l0rd) [reply infrastructures/kubernetes/pom.xml]: @skabashnyuk the same applies to the `docker-environment` dependency that is just below right?

**Review by l0rd — APPROVED** [review]

File: infrastructures/kubernetes/pom.xml (hunk 8 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/pom.xml]: I would say that not the same. It is supposed that `docker-environment` may be reused by another infrastructure.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+import javax.inject.Inject;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.infrastructure.docker.client.UserSpecificDockerRegistryCredentialsProvider;
+import org.eclipse.che.infrastructure.docker.client.dto.AuthConfigs;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespaceFactory;
+
+/** @author David Festal */
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: Fixed

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    this.clientFactory = clientFactory;
+    this.namespaceFactory = namespaceFactory;
+  }
+
+  @Override
+  public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+
+    AuthConfigs authConfigs = userSpecificDockerRegistryCredentialsProvider.getCredentials();
+    if (authConfigs.getConfigs().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: Fixed

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidfestal) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: Fixed with a private method

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        JsonWriter jsonWriter = gson.newJsonWriter(strWriter)) {
+      jsonWriter.beginObject();
+
+      authConfigs
+          .getConfigs()
+          .forEach(
+              (name, authConfig) -> {
+                try {
+                  if (!name.startsWith("https://")) {
+                    name = "https://" + name;
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: Fixed

**Comment by davidfestal** [comment]
@garagatyi @sleshchenko  I mainly did the changes you requested. Is it OK for you OK now ?

@skabashnyuk The only remaining comment that I didn't fix is [this one](https://github.com/eclipse/che/pull/9962#pullrequestreview-127063021)

@l0rd wdyt ? could be postpone this to a next PR ?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespaceFactory;
+
+/** @author David Festal */
+public class ImagePullSecretProvisioner implements ConfigurationProvisioner<KubernetesEnvironment> {
+
+  static final String SECRET_NAME = "workspace-private-registries";
+
+  private final UserSpecificDockerRegistryCredentialsProvider
+      userSpecificDockerRegistryCredentialsProvider;
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: tiny one. I don't like such long names :smile:

File: infrastructures/kubernetes/pom.xml (hunk 8 above)
Comment (skabashnyuk) [reply infrastructures/kubernetes/pom.xml]: Yes. Could you do that please in this PR. I really appreciate your efforts. I guess it will take less time then we spend writing this comments and protect us with potential binding/classloading/configuration issues.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 14]
```diff
@@ -127,7 +109,73 @@ public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
...
+          String auth =
+              new StringBuilder(authConfig.getUsername())
+                  .append(':')
+                  .append(authConfig.getPassword())
+                  .toString();
+          jsonWriter.name("auth");
+          jsonWriter.value(encoder.encodeToString(auth.getBytes()));
+          jsonWriter.endObject();
+        } catch (IOException e) {
+          throw new InfrastructureException(e);
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: I think that IOException message here won't be user-friendly. Can you wrap it with an introduction appropriate for a user?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java [hunk 15]
```diff
@@ -127,7 +109,73 @@ public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
...
+        }
+      }
+
+      jsonWriter.endObject();
+      jsonWriter.flush();
+      return strWriter.toString();
+    } catch (IOException e) {
+      throw new InfrastructureException(e);
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java:177]: I think that IOException message here won't be user-friendly. Can you wrap it with an introduction appropriate for a user?

**Review by garagatyi — APPROVED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (davidfestal) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java]: I reduced its length :-D

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java (hunk 15 above)
Comment (davidfestal) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java:177]: I added a message to the wrapping exception.

File: infrastructures/kubernetes/pom.xml (hunk 8 above)
Comment (davidfestal) [reply infrastructures/kubernetes/pom.xml]: I extracted the docker auth-related classes to a new `docker-auth` maven module.
@skabashnyuk does it correspond to what you were expecting ?

**Comment by davidfestal** [comment]
ci-test

**Review by skabashnyuk — APPROVED** [review]

**Review by ibuziuk — APPROVED** [review]

File: infrastructures/docker/docker-auth/pom.xml [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        <artifactId>che-infrastructures-docker-parent</artifactId>
+        <groupId>org.eclipse.che.infrastructure.docker</groupId>
+        <version>6.7.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <artifactId>docker-auth</artifactId>
+    <packaging>jar</packaging>
+    <name>Infrastructure :: Docker :: Docker Auth</name>
+    <properties>
+        <findbugs.failonerror>false</findbugs.failonerror>
```
Comment (sleshchenko) [inline infrastructures/docker/docker-auth/pom.xml]: Can it be removed?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/convert/DockerComposeEnvironmentConverter.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+
+/**
+ * Converts {@link DockerImageEnvironment} to {@link KubernetesEnvironment}.
+ *
+ * @author Sergii Leshchenko
+ * @author Anton Korneta
+ */
+@Singleton
+public class DockerComposeEnvironmentConverter {
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/convert/DockerComposeEnvironmentConverter.java]: It looks exactly the same as `DockerImageEnvironmentConverter` but with another name. Please check these changes, I guess it should be reverted.

**Review by sleshchenko — COMMENTED** [review]
LGTM. Please take a look my inlined comments especially about DockerComposeEnvironmentConverter.

File: infrastructures/docker/docker-auth/pom.xml (hunk 16 above)
Comment (garagatyi) [reply infrastructures/docker/docker-auth/pom.xml]: It should be! Please, do not use this property unless you have strong arguments why we should use it!

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/convert/DockerComposeEnvironmentConverter.java (hunk 17 above)
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/convert/DockerComposeEnvironmentConverter.java]: This class seems to be unrelated to your PR. Can you elaborate why we need it?

**Review by garagatyi — CHANGES_REQUESTED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/convert/DockerComposeEnvironmentConverter.java (hunk 17 above)
Comment (davidfestal) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/convert/DockerComposeEnvironmentConverter.java]: yes, it was added by mistake. I just removed it.

File: infrastructures/docker/docker-auth/pom.xml (hunk 16 above)
Comment (davidfestal) [reply infrastructures/docker/docker-auth/pom.xml]: This came from a copy/paste of the `docker-client` `pom.xml` file. 
I just removed it.

**Comment by davidfestal** [comment]
ci-test

**Review by sleshchenko — APPROVED** [review]

**Review by garagatyi — APPROVED** [review]
LGTM
